### ISSUE TITLE
20.3(Nexus)  no longer map Function keys on macOS

### ISSUE THREAD

**Issue by xbmute** [xbmc/xbmc#24544]
### Describe the bug

mapping of Function keys (F1 - F12) using keymaps/keyboard.xml no longer registers after upgrading
from 20.2 to 20.3 on macOS. This behavior also exist in the Omega(21.x) beta series and master nightlies.

## Expected Behavior

pressing the Function key leading to executing the mapped kodi Action.

## Actual Behavior

Kodi does not register function keys when pressed.

## Possible Fix

Unknown

### To Reproduce

1. add the following xml to the keymaps folder
```
    <keymap>
      <global>
        <keyboard>
            <f1>ActivateWindow(Music,Files)</f1>
            <f2>ActivateWindow(Videos,Files)</f2>
            <f3>ActivateWindow(home)</f3>
	</keyboard>
      </global>
    </keymap>
```
2. restart kodi
3. use the mapped F1, F2, F3 to switch to the window specified

### Debuglog

The debuglog can be found here: see attachments

### Screenshots 
Here are some links or screenshots to help explain the problem:

Not Needed

## Additional context or screenshots (if appropriate)

None

### Your Environment
Used Operating system:

 - [ ] Android
 - [ ] iOS
 - [ ] tvOS
 - [ ] Linux
 - [x] macOS
 - [ ] Windows
 - [ ] Windows UWP

 - Operating system version/name: macOS 12.7.2 (Monterey), also tested on 10.15.7 (Catalina)
 - Kodi version: 20.3

*note: Once the issue is made we require you to update it with new information or Kodi versions should that be required.
Team Kodi will consider your problem report however, we will not make any promises the problem will be solved.*

**Comment by xbmute** [comment]
attached debug logs for 20.3(not functioning) and 20.2(working):
[kodi.log-no-functionkey-20.3.txt](https://github.com/xbmc/xbmc/files/13992253/kodi.log-no-functionkey-20.3.txt)
[kodi.log-function-key-working-20.2.txt](https://github.com/xbmc/xbmc/files/13992255/kodi.log-function-key-working-20.2.txt)

**Comment by kambala-decapitator** [comment]
can reproduce on Sonoma with 20.3

I suspect this issue is related https://forum.kodi.tv/showthread.php?tid=375805 - when pressing PageUp I also see `HandleKey: 0 (0xf200, obc-61697) pressed` just like with Fx

**Comment by xbmute** [comment]
digging around,  it appears that we are no longer using xbmc/input/ButtonTranslater.cpp and family anymore, was garbear who build the new replacement input system?

**Comment by kambala-decapitator** [comment]
@garbear any idea?

**Comment by enen92** [comment]
On v21 with the move to native windowing Kodi does not even catch such events. Those are mostly the type NSSystemDefined (just like media keys) and are sort of reserved...
For instance F1 and F2 change the screen brightness on my Mac should we actually grab those keys? Even catching them won't avoid the system to change the brightness first.
I am pretty tempted to simply mark this as won't fix...
Can't you simply use other keys for those shortcuts?

**Comment by kambala-decapitator** [comment]
> For instance F1 and F2 change the screen brightness on my Mac

@enen92  the issue is about Fn+F1 etc., it's not about those system functions. I use a PC keyboard that has dedicated Fx keys as well as various media keys.

also, it's reproducible in Nexus, and see https://forum.kodi.tv/showthread.php?tid=375805 - the report looks almost the same, but it's about Page Up and related keys.

**Comment by garbear** [comment]
For keyboards, mice, remotes, event clients, everything but joysticks, the new input stuff is built as a layer on top of the existing code, which hasn't changed AFAIK.

If it's windowing that caused the problem, it could be one of these PRs: https://github.com/xbmc/xbmc/issues?q=is%3Aclosed+milestone%3A%22Nexus+20.3%22+label%3A%22Platform%3A+macOS%22

If input is the problem, here's the input PRs in v20.3, though notice that they're all mine and someone else may have forgotten to mark the input component tag: https://github.com/xbmc/xbmc/issues?q=is%3Aclosed+milestone%3A%22Nexus+20.3%22+label%3A%22Component%3A+Input%22

**Comment by xbmute** [comment]
this bug does not appear to trigger on Windows, tested on Windows 10 Pro (22H2), in both cases the Function keys behaved as expected.

**Comment by xbmute** [comment]
I can attempt some bisect testing to narrow down the offending commit. I see that Nexus nightly dmgs are only kept for 1 month on the mirror. Is there an archive where I can locate mac nightly builds between Jun 30, 2023 - Oct 17, 2023?

**Comment by thanosz** [comment]
I also have the same problems - all default key mapping seem to not work - i.e. Mute with fn+F8, vol up/down with fn+F9/fn+F10, etc

**Comment by kambala-decapitator** [comment]
I can confirm that F1, PgUp and similar buttons work in 20.2 on macOS, but neither in 20.3 nor in Omega

also just tried to use the keymap editor addon to assign F1 and F2 in Omega and result is the same button 0xf200, also see my comment above https://github.com/xbmc/xbmc/issues/24544#issuecomment-1901895260

![Screenshot 2024-01-21 at 12 08 07](https://github.com/xbmc/xbmc/assets/1557784/b76f9873-dbb4-4bf8-a86a-2a3eb7a9d1ac)

**Comment by kambala-decapitator** [comment]
I'm pretty sure it's not Windowing because:
1. there were no macOS windowing changes between 20.2 and 20.3
2. Nexus and Omega use different Windowing implementation

None of the PRs from https://github.com/xbmc/xbmc/issues?q=is%3Aclosed+milestone%3A%22Nexus+20.3%22+label%3A%22Platform%3A+macOS%22 are related to this issue

**Comment by enen92** [comment]
> For keyboards, mice, remotes, event clients, everything but joysticks, the new input stuff is built as a layer on top of the existing code, which hasn't changed AFAIK.
> 
> If it's windowing that caused the problem, it could be one of these PRs: https://github.com/xbmc/xbmc/issues?q=is%3Aclosed+milestone%3A%22Nexus+20.3%22+label%3A%22Platform%3A+macOS%22
> 
> If input is the problem, here's the input PRs in v20.3, though notice that they're all mine and someone else may have forgotten to mark the input component tag: https://github.com/xbmc/xbmc/issues?q=is%3Aclosed+milestone%3A%22Nexus+20.3%22+label%3A%22Component%3A+Input%22

Just gave it a go using keymap editor addon on current master on macOS and pressing fn+F1 to edit a given action, results on the following file:

```
<keymap><global><keyboard><key id="61952">runaddon(script.keymap)</key></keyboard></global></keymap>
```

Pressing fn+F1 correctly opens the add-on, so the shortcut is well interpreted. So the issue is that `<f1>` and similar keys are not being mapped to the proper key ids?

**Comment by kambala-decapitator** [comment]
@enen92 all Fx (fn+Fx on Apple keyboards) buttons as well as PgUp (fn+Up) and similar map to a single 61952 value, see my comment/SS above. Navigation with PgUp and similar no longer works, arrows still work.

`<f1>` issue looks tightly related

**Comment by enen92** [comment]
This works for me in master using the keymap xml file provided in the bug report:

```
From 1486f4b8efee4f1a7d0f6724712c913228535af0 Mon Sep 17 00:00:00 2001
From: Miguel Borges de Freitas <92enen@gmail.com>
Date: Sun, 21 Jan 2024 10:05:32 +0000
Subject: [PATCH] MacOS: add fx and pageup/down mappings

---
 xbmc/windowing/osx/WinEventsOSXImpl.mm | 32 ++++++++++++++++++++++++++
 1 file changed, 32 insertions(+)

diff --git a/xbmc/windowing/osx/WinEventsOSXImpl.mm b/xbmc/windowing/osx/WinEventsOSXImpl.mm
index 4a35ba1c20..c1261837d0 100644
--- a/xbmc/windowing/osx/WinEventsOSXImpl.mm
+++ b/xbmc/windowing/osx/WinEventsOSXImpl.mm
@@ -101,8 +101,10 @@ - (unichar)OsxKey2XbmcKey:(unichar)character
     case NSRightArrowFunctionKey:
       return XBMCK_RIGHT;
     case NSUpArrowFunctionKey:
+    case NSPageUpFunctionKey:
       return XBMCK_UP;
     case NSDownArrowFunctionKey:
+    case NSPageDownFunctionKey:
       return XBMCK_DOWN;
     case NSBackspaceCharacter:
     case NSDeleteCharacter:
@@ -110,6 +112,36 @@ - (unichar)OsxKey2XbmcKey:(unichar)character
     case NSCarriageReturnCharacter:
     case NSEnterCharacter:
       return XBMCK_RETURN;
+    case NSF1FunctionKey:
+      return XBMCK_F1;
+    case NSF2FunctionKey:
+      return XBMCK_F2;
+    case NSF3FunctionKey:
+      return XBMCK_F3;
+    case NSF4FunctionKey:
+      return XBMCK_F4;
+    case NSF5FunctionKey:
+      return XBMCK_F5;
+    case NSF6FunctionKey:
+      return XBMCK_F6;
+    case NSF7FunctionKey:
+      return XBMCK_F7;
+    case NSF8FunctionKey:
+      return XBMCK_F8;
+    case NSF9FunctionKey:
+      return XBMCK_F9;
+    case NSF10FunctionKey:
+      return XBMCK_F10;
+    case NSF11FunctionKey:
+      return XBMCK_F11;
+    case NSF12FunctionKey:
+      return XBMCK_F12;
+    case NSF13FunctionKey:
+      return XBMCK_F13;
+    case NSF14FunctionKey:
+      return XBMCK_F14;
+    case NSF15FunctionKey:
+      return XBMCK_F15;
     default:
       return character;
   }
--
2.39.3 (Apple Git-145)
```

Also added pageup and pagedown, is this what is expected to happen?

**Comment by kambala-decapitator** [comment]
also home/end and insert are missing ;)

but it doesn't explain how it broke in 20.3

**Comment by enen92** [comment]
To what keys should them map?

**Comment by kambala-decapitator** [comment]
> To what keys should them map?

well, to the same keys, see here: https://github.com/xbmc/xbmc/blob/master/xbmc/input/keyboard/XBMC_keytable.cpp#L217

**Comment by enen92** [comment]
This has been fixed for omega, keeping it open because of the nexus regression (not sure anyone will look into it nor even if there'll be another nexus release though)

**Comment by xbmute** [comment]
Thank you! I will try upgrading to tomorrow's Nexus nightly build.

**Comment by xbmute** [comment]
@enen92, may we have this merged in the Nexus branch also, so those of us may benefit by using the Nexus nighty builds?

**Comment by kambala-decapitator** [comment]
> @enen92, may we have this merged in the Nexus branch also, so those of us may benefit by using the Nexus nighty builds?

that solution is strictly for Omega - Nexus requires something else

**Comment by xbmute** [comment]
Ahhh, thank you for clarifying.  I see now  how they are quite different.  Would you kindly offer some implementation pointers, so I can possibly work towards a PR for Nexus?

**Comment by kambala-decapitator** [comment]
> Ahhh, thank you for clarifying. I see now how they are quite different. Would you kindly offer some implementation pointers, so I can possibly work towards a PR for Nexus?

sorry, currently no idea what could break in 20.3

**Comment by xbmute** [comment]
Thank you for fixing this in master. Happy New Year and keep up the good work!

**Comment by kambala-decapitator** [comment]
built debug version of Nexus locally and all buttons works fine... Now that's super weird...

**Comment by kambala-decapitator** [comment]
in my local version I didn't build any addons. Just checked the latest nightly - buttons don't work.

log from my version:
```
Keyboard: scancode: 0x79, sym: 0x119, unicode: 0xf72d, modifier: 0x0
HandleKey: pagedown (0xf085) pressed, window 10000, action is PageDown
Keyboard: scancode: 0x79, sym: 0x119, unicode: 0x00, modifier: 0x0
```

from nightly:
```
Keyboard: scancode: 0x79, sym: 0x0c, unicode: 0x0c, modifier: 0x0
GetActionCode: Trying Hardy keycode for 0xf200
Skipped 3 duplicate messages..
HandleKey: 0 (0xf200, obc-61697) pressed, window 10000, action is 
```

does it mean that some addon is involved?

**Comment by garbear** [comment]
peripheral.joystick isn't used for keyboard translation but it might be worth seeing if it's the cause.

**Comment by kambala-decapitator** [comment]
> peripheral.joystick isn't used for keyboard translation but it might be worth seeing if it's the cause.

doesn't look so

**Comment by xbmute** [comment]
those are the only addons here:  peripheral.joystick, screensaver.picture.slideshow, script.skinshortcuts, skin.estuary, skin.unity.  

Trying to setup a local env to build. docs/README.macOS.md says it is using 10.14(Line 4) and then 10.15.7(Line 62). What version of macOS/Xcode/SDK are the official builds made on?

**Comment by kambala-decapitator** [comment]
> Trying to setup a local env to build. docs/README.macOS.md says it is using 10.14(Line 4) and then 10.15.7(Line 62). What version of macOS/Xcode/SDK are the official builds made on?

environment is: https://github.com/xbmc/xbmc/blob/Nexus/tools/buildsteps/defaultenv#L45-L48

build machine is running macOS Ventura (13.5.2), but that shouldn't matter much

I'm building on Sonoma with Xcode 14.2

**Comment by kambala-decapitator** [comment]
I've found the reason: Jenkins builds with `APP_WINDOW_SYSTEM=native` instead of `sdl`

**Comment by kambala-decapitator** [comment]
fixed on Jenkins, all next nightlies will be fixed. You can also grab the currently fixed version here (once the build is finished): https://jenkins.kodi.tv/job/OSX-64/33062/

looks like the reason was me pasting updated notarization script without checking which directory build steps are read from, sorry for that.

**Comment by xbmute** [comment]
@kambala-decapitator, thank you for getting to the bottom of this!

**Comment by liquidcms** [comment]
Is this just a macos issue? My kep mappings (all of them) on my nvidia shield suddenly stopped working. I suspect it possibly did an update on its own; although can't say for sure. It was Nexus, and now it is Nexus 20.3. I upgraded to 20.5 and doesnt fix the issue.

I tried to downgrade to 20.2 but it fails. I suspect i will need to uninstall completely and re-install 20.2 (and disable auto update, assuming that's what happened).

The keys are recognized as i can set the mapping; they just aren't used.

**Comment by garbear** [comment]
@liquidcms can you upload a debug log? I've been looking into Android input issues

**Comment by liquidcms** [comment]
not the entire log, but the error created when trying to use a mapped key:

2024-03-12 00:01:58.979 T:4685     info <general>: Loading skin file: DialogFavourites.xml, load type: KEEP_IN_MEMORY
2024-03-12 00:01:58.980 T:4685    error <general>: Unable to load window XML: /storage/emulated/0/Android/data/org.xbmc.kodi/files/.kodi/addons/skin.confluence/720p/DialogFavourites.xml. Line 0
                                                   Failed to open file

**Comment by garbear** [comment]
Maybe you should provide a full log, because for one the beginning is critical as we fully log all input devices and how they look to Android, and second your log snippet shows a skin error not an input error.

**Comment by kambala-decapitator** [comment]
This issue was only about macos.

@liquidcms please open a separate issue
